pkg/kds/mux/zone_sync.go

```diff
@@ -144,18 +179,74 @@ func (g *KDSSyncServiceServer) ZoneToGlobalSync(stream mesh_proto.KDSSyncService
 	}
 }
 
-func (g *KDSSyncServiceServer) watchZoneHealthCheck(streamContext context.Context, zone string) events.Listener {
+func (g *KDSSyncServiceServer) watchZoneHealthCheck(streamContext context.Context, zone string, typ service.StreamType, connectTime time.Time) events.Listener {
 	tenantID, _ := multitenant.TenantFromCtx(streamContext)
 
 	shouldDisconnectStream := events.NewNeverListener()
-
 	if kds.ContextHasFeature(streamContext, kds.FeatureZonePingHealth) {
 		shouldDisconnectStream = g.eventBus.Subscribe(func(e events.Event) bool {
-			disconnectEvent, ok := e.(service.ZoneWentOffline)
-			return ok && disconnectEvent.TenantID == tenantID && disconnectEvent.Zone == zone
+			switch event := e.(type) {
+			case service.ZoneWentOffline:
+				return event.TenantID == tenantID && event.Zone == zone
+			case service.StreamCancelled:
+				return event.TenantID == tenantID && event.Zone == zone && event.Type == typ && event.ConnTime == connectTime
+			default:
+				return false
+			}
 		})
-		g.eventBus.Send(service.ZoneOpenedStream{Zone: zone, TenantID: tenantID})
+		g.eventBus.Send(service.ZoneOpenedStream{Zone: zone, TenantID: tenantID, Type: typ, ConnTime: connectTime})
 	}
 
 	return shouldDisconnectStream
 }
+
+func (g *KDSSyncServiceServer) storeStreamConnection(ctx context.Context, zone string, typ service.StreamType, connectTime time.Time) error {
+	key := model.ResourceKey{Name: zone}
+
+	// wait for Zone to be created, only then we can create Zone Insight
+	err := retry.Do(
+		ctx,
+		retry.WithMaxRetries(30, retry.NewConstant(1*time.Second)),
+		func(ctx context.Context) error {
+			return retry.RetryableError(g.resManager.Get(ctx, system.NewZoneResource(), core_store.GetBy(key)))
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Add delay for Upsert. If Global CP is behind an HTTP load balancer,
+	// it might be the case that each Envoy Admin stream will land on separate instance.
+	// In this case, all instances will try to update Zone Insight which will result in conflicts.
+	// Since it's unusual to immediately execute envoy admin rpcs after zone is connected, 0-10s delay should be fine.
+	// #nosec G404 - math rand is enough
+	time.Sleep(time.Duration(rand.Int31n(10000)) * time.Millisecond)
```

Comment:
Isn't the retry below just fine?

Reply:
It's fine, but it doesn't solve the issue with the first request. It only improves the probability that the first request won't hit a conflict.